File: Solutions/653. Two Sum IV - Input is a BST/653. Two Sum IV - Input is a BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 
 //set implementation
class Solution {
public:
    unordered_set<int> s;
    bool findTarget(TreeNode* root, int k) {
        if(!root)
            return false;
        if(s.count(k-root->val))
            return true;
        s.insert(root->val);
        return findTarget(root->left,k) or findTarget(root->right,k);
    }
};



//binary search method
class Solution {
public:
    TreeNode *root;
    bool findTarget(TreeNode* root, int k) {
        if(!this->root)
            this->root=root;
        if(!root)
            return false;
        if(search(root,k-root->val))
            return true;
        return findTarget(root->left,k) or findTarget(root->right,k);
    }
    bool search(TreeNode* root,int complement){
        TreeNode* cur=this->root;
        while(cur){
            if(complement>cur->val)
                cur=cur->right;
            else if(complement<cur->val)
                cur=cur->left;
            else
                return cur==root?false:true;
        }
        return false;
    }
};

File: Solutions/653. Two Sum IV - Input is a BST/653. Two Sum IV - Input is a BST.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 
//set implementation
func findTarget(root *TreeNode, k int) bool {
    set:=make(map[int]bool)
    return dfs(root,set,k)
}
func dfs(root *TreeNode,set map[int]bool,k int) bool{
    if(root==nil){
        return false;
    }
    if _,ok:=set[k-root.Val];ok{
        return true;
    }
    set[root.Val]=true
    return dfs(root.Left,set,k) ||  dfs(root.Right,set,k)
}


//binary search method

func foo(root *TreeNode,Begin *TreeNode,k int) bool{
    if Begin==nil{
        Begin=root
    }
    if root==nil{
        return false
    }
    if search(root,k-root.Val,Begin){
        return true
    }
    return foo(root.Left,Begin,k) || foo(root.Right,Begin,k)
}

func findTarget(root *TreeNode, k int) bool {
    var Begin *TreeNode=nil
    return foo(root,Begin,k)
}
func search(root *TreeNode,complement int,Begin *TreeNode) bool{
    var curr *TreeNode=Begin
    for curr!=nil{
        if(curr.Val>complement){
            curr=curr.Left
        }else if(curr.Val<complement){
            curr=curr.Right
        }else{
            if curr==root{
                return false
            }
            return true
        }
    }
    return false
}